src/subscription.ts

```diff
@@ -1,263 +1,71 @@
-            return new Error("Invalid event ID in message: " + JSON.stringify(eventMessage));
-        }
-        if (typeof headers !== "object" || Array.isArray(headers)) {
-            return new Error("Invalid event headers in message: " + JSON.stringify(eventMessage));
-        }
-
-        if (this.options.onEvent) { this.options.onEvent({ eventId: id, headers: headers, body: body }); }
-    }
-
-    // calls options.onEvent 0+ times, then possibly returns an error
-    private onEOSMessage(eosMessage: any[]): Error {
-        this.assertState(['OPEN']);
+export interface SubscriptionState {
+    unsubscribe();
+}
 
-        if (eosMessage.length !== 4) {
-            return new Error("EOS message has " + eosMessage.length + " elements (expected 4)");
-        }
-        let [_, statusCode, headers, info] = eosMessage;
-        if (typeof statusCode !== "number") {
-            return new Error("Invalid EOS Status Code");
-        }
-        if (typeof headers !== "object" || Array.isArray(headers)) {
-            return new Error("Invalid EOS Headers");
-        }
+export interface SubscriptionEvent {
+    eventId: string;
+    headers: ElementsHeaders;
+    body: any;
+}
 
-        this.state = SubscriptionState.ENDING;
-        return new ErrorResponse(statusCode, headers, info);
-    }
+export interface SubscriptionListeners {
+    onOpen?: (headers: ElementsHeaders) => void;
+    onSubscribe?: () => void;
+    onRetrying?:() => void;
+    onResuming?:() => void;
+    onEvent?: (event: SubscriptionEvent) => void;
+    onError?: (error: any) => void;
+    onEnd?: (error: any) => void;
+}
 
-    unsubscribe(err?: Error) {
-        this.state = SubscriptionState.ENDED;
-        this.xhr.abort();
-        if (err) {
-            if (this.options.onError) { this.options.onError(err); }
-        } else {
-            if (this.options.onEnd) { this.options.onEnd(); }
-        }
+let noop = (arg?) => {};
+
+export let replaceMissingListenersWithNoOps: (listeners: SubscriptionListeners) => SubscriptionListeners = (listeners) => {
+    
+    let onOpen = listeners.onOpen || noop ;
+    let onSubscribe = listeners.onSubscribe || noop;
+    let onEvent = listeners.onEvent || noop;
+    let onError = listeners.onError || noop;
+    let onEnd = listeners.onEnd || noop; 
+    let onRetrying = listeners.onRetrying || noop;
+    let onResuming = listeners.onResuming || noop;
+
+    return {
+        onOpen: onOpen,
+        onSubscribe: onSubscribe,
+        onRetrying: onRetrying,
+        onResuming: onResuming,
+        onEvent: onEvent,
+        onError: onError,
+        onEnd: onEnd
     }
 }
+
+export type SubscriptionConstructor = (
+    onOpen: (headers:ElementsHeaders) => void , 
+    onError: (error: any) => void, 
+    onEvent: (event: SubscriptionEvent) => void,
+    onEnd: (error: any) => void,
+    headers: ElementsHeaders, 
+) => BaseSubscription;
+
+export type SubscribeStrategy = (
+    onOpen: (headers:ElementsHeaders) => void, 
```

Comment:
Could the params in this function be grouped as an interface?

Otherwise reading code like this becomes hard(er) (for me at least!):

```js
let executeNextSubscribeStrategy = () => {
  logger.verbose(`RetryingSubscription: trying to re-establish the subscription`);
 
  let underlyingSubscription = nextSubscribeStrategy(headers => {
    onTransition(new OpenSubscriptionState(headers, underlyingSubscription, onTransition));
  },
  onRetrying,
  error => {
    executeSubscriptionOnce(error);
  }, 
  onEvent, 
  error => {
    onTransition(new EndedSubscriptionState(error));
  },
  headers, 
  subscriptionConstructor);
};
```